ooni/resources.py

```diff
@@ -0,0 +1,179 @@
+            #  the manifest claims we have a more up to date version.
+            # This happens if an update by country_code happened and a new
+            # country code is now required.
+            if filename.endswith(".gz"):
+                filename = filename[:-3]
+            if not _resources.child(pre_path).child(filename).exists():
+                paths_to_update.append(info)
+
+    return paths_to_update, paths_to_delete
+
+
+@defer.inlineCallbacks
+def check_for_update(country_code=None):
+    """
+    Checks if we need to update the resources.
+    If the country_code is specified then only the resources for that
+    country will be updated/downloaded.
+    :return: the latest version.
+    """
+    temporary_files = []
+    def cleanup():
+        # If we fail we need to delete all the temporary files
+        for _, src_file_path in temporary_files:
+            src_file_path.remove()
+
+    current_version = get_current_version()
+    latest_version = yield get_latest_version()
+
+    resources_dir = FilePath(config.resources_directory)
+    try:
+        resources_dir.makedirs()
+    except OSError as e:
+        if not e.errno == errno.EEXIST:
+            raise
+    current_manifest = resources_dir.child("manifest.json")
+
+    if current_manifest.exists():
+        with current_manifest.open("r") as f:
+            current_manifest_data = json.load(f)
+    else:
+        current_manifest_data = {
+            "resources": []
+        }
+
+    # We should download a newer manifest
+    if current_version < latest_version:
+        new_manifest = current_manifest.temporarySibling()
+        new_manifest.alwaysCreate = 0
+
+        temporary_files.append((current_manifest, new_manifest))
+
+        try:
+            yield downloadPage(
+                get_download_url(latest_version, "manifest.json"),
+                new_manifest.path
+            )
+        except:
+            cleanup()
+            raise UpdateFailure("Failed to download manifest")
+
+        new_manifest_data = json.loads(new_manifest.getContent())
+    else:
+        new_manifest_data = current_manifest_data
+
+    to_update, to_delete = get_out_of_date_resources(
+            current_manifest_data, new_manifest_data, country_code)
+
+    try:
+        for resource in to_update:
+            gzipped = False
+            pre_path, filename = resource["path"].split("/")
+            if filename.endswith(".gz"):
+                filename = filename[:-3]
+                gzipped = True
+            dst_file = resources_dir.child(pre_path).child(filename)
+            try:
+                dst_file.parent().makedirs()
+            except OSError as e:
+                if not e.errno == errno.EEXIST:
+                    raise
+            src_file = dst_file.temporarySibling()
+            src_file.alwaysCreate = 0
+
+            temporary_files.append((dst_file, src_file))
+            # The paths for the download require replacing "/" with "."
```

Comment:
I think this part of the protocol is quite fragile. What about specifying the desired directory relative to the resources folder separately from the file name.